components/button/floating-buttons.js

```diff
@@ -99,32 +101,48 @@ class FloatingButtons extends RtlMixin(LitElement) {
 	constructor() {
 		super();
 		this.alwaysFloat = false;
-		this._calcContainerPosition = this._calcContainerPosition.bind(this);
+		this._containerMarginLeft = '';
+		this._containerMarginRight = '';
+		this._floating = false;
+		this._innerContainerLeft = '';
+		this._innerContainerRight = '';
+		this._intersectionObserver = null;
+		this._isIntersecting = false;
+		this._recalculateFloating = this._recalculateFloating.bind(this);
 	}
 
 	connectedCallback() {
 		super.connectedCallback();
-		window.addEventListener('scroll', this._calcContainerPosition);
-		window.addEventListener('resize', this._calcContainerPosition);
-		window.addEventListener('d2l-dir-update', this._calcContainerPosition);
+		if (mediaQueryList.addEventListener) mediaQueryList.addEventListener('change', this._recalculateFloating);
 	}
 
 	disconnectedCallback() {
 		super.disconnectedCallback();
-		window.removeEventListener('scroll', this._calcContainerPosition);
-		window.removeEventListener('resize', this._calcContainerPosition);
-		window.removeEventListener('d2l-dir-update', this._calcContainerPosition);
-		if (this.__resizeObserver) {
-			this.__resizeObserver.disconnect();
-			this.__resizeObserver = null;
-		}
+		if (mediaQueryList.removeEventListener) mediaQueryList.removeEventListener('change', this._recalculateFloating);
+		if (this._intersectionObserver) this._intersectionObserver.disconnect();
 	}
 
 	firstUpdated() {
 		super.firstUpdated();
 
-		this._calcContainerPosition();
-		this._startObserver();
+		const test = document.createElement('div');
```

Comment:
If I understand this correctly, we need something outside this component to test, that doesn't get floated...?

In the code below, is there only ever one entry?  Just wondering if there is more than one entry, are we calling the recalc function more than once?

Reply:
Right, I meant to put a note in about this. I was originally hoping this fake element could just be output in `render()` and the styles that are getting applied to the host could move to an inner container.

The problem is with sticky positioning -- the way it works is that the parent element always becomes the sticky container and the siblings are the things it "sticks" against. So the host HAS to be the thing that's sticky, so that its parent is the sticky container. That meant manually inserting the fake element as the host's next sibling.

In terms of the observer, you end up with one entry for each target element that you're observing. Since we're only observing 1 element (the fake test element), we'll only ever get 1 entry.

Reply:
I see, makes sense.  So in this case, if the floating buttons are moved in the DOM (I think unlikely), this would currently stop working.  Also, if the app inserts anything between the test node and the custom element, we could get some odd behaviour.  Not sure how much we care about these scenarios.

Reply:
Correct. But the moving thing is a valid concern... I'll see if it's easy to handle on disconnect/connect.

Reply:
I just pushed some changes that handle removing/re-adding the test element when the buttons are disconnected and re-connected.

I also noticed that on really slow pages, the floating buttons would be hidden (and therefore float), then be visible (and un-float), then get pushed down as content loaded and become floating again... which lead to 3-4 re-calculations. So I've moved the call to `observe` into a `requestIdleCallback` with a 5 second timeout. Having the buttons immediately float just isn't that important so I think deferring a bit is fine.